crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -2240,17 +2247,15 @@ where
                 ));
             } else {
                 self.deferred
-                    .type_definitions
+                    .future_type_definitions
                     .push((expr, self.ctx.snapshot()));
             }
             return;
         }
 
         self.ctx.push_expr(expr);
 
-        let prev_in_literal = self.ctx.in_literal;
-        let prev_in_type_definition = self.ctx.in_type_definition;
-        let prev_in_boolean_test = self.ctx.in_boolean_test;
+        let flags = self.ctx.flags;
```

Comment:
So the new flow is: we save the flags, then recurse and set any necessary flags, then restore the previous flags when we finish the recursion phase (end of `visit_expr`).